api-update/Jenkinsfile-deploy

```diff
@@ -0,0 +1,44 @@
+pipeline {
+    agent {
+        label "master"
+    }
+    parameters {
+        choice(name: "STAGE", choices: ["intg", "staging", "prod"], description: "The stage you are building the front end for")
+        string(name: "TO_DEPLOY", description: "The git tag, branch or commit reference to deploy, e.g. 'v123'")
+    }
+    stages {
+        stage("Deploy lambda") {
+            agent {
+                ecs {
+                    inheritFrom "aws"
+                    taskrole "arn:aws:iam::${env.MANAGEMENT_ACCOUNT}:role/TDRJenkinsNodeLambdaRole${params.STAGE.capitalize()}"
+                }
+            }
+            steps {
+                script {
+                    def accountNumber = getAccountNumberFromStage()
+                    VERSION_TAG = sh (
+                        script: "python3 /deploy_lambda_from_s3.py ${accountNumber} ${params.STAGE} tdr-api-update-antivirus-${params.STAGE} tdr-backend-checks-${params.STAGE} antivirus.jar",
+                        returnStdout: true
+                    ).trim()
+                    sh 'git config --global user.email tna-digital-archiving-jenkins@nationalarchives.gov.uk'
+                    sh 'git config --global user.name tna-digital-archiving-jenkins'
+                    sh "git tag ${VERSION_TAG}"
+                    sshagent(['github-jenkins']) {
+                        sh("git push origin ${VERSION_TAG}")
```

Comment:
You're right, this is a bit backwards. I've changed it so now it goes like:
**Jenkinsfile-testing**
* Deploys code to a new version, the version number is sequential and provided by lambda. 
* Tags the branch with this version.
* Triggers the deploy job, passing the version (and the stage) as a parameter

**Jenkinsfile-deploy**
* Updates the mapping between the sqs queue and the lambda to use the version passed in the arguments.

This means that we can deploy any particular version on any particular environment by moving the mapping between the sqs queue and the lambda.

Reply:
This also needed a change to the jenkins python scripts which is here if you want to see it.
https://github.com/nationalarchives/tdr-jenkins/pull/53

Reply:
Sorry, just thought of another potential issue: will this work cross-environment? Won't it end up with a different version tag on intg/staging/prod?

Reply:
Yeah, it will end up with different versions but I can't think of a clean way to do it. To deploy to staging, we'd check out the version we want to deploy from git, build it, update the staging lambda and then run the Jenkins job with that version. 
This isn't too difficult. If you've got any easily automated solutions then I can put those in but if it's something we need to think about later then it would be better to get this in as is as it's blocking me from doing the rest of the backend checks properly.

Reply:
Which Jenkins job does the "check out the version we want to deploy from git, build it, update the staging lambda" bit?

Reply:
There isn't one at the moment. I think it can be done in the deployment job. The order you have to do things is
1. Deploy the code to $LATEST
2. Update the lambda version, getting the version number in the response.
3. Update the sqs mapping.

For the integration environment, the first two steps are done in Jenkinsfile-testing which works but you don't want to be updating the staging and prod code every time we do the integration one so you can't do those environments in there. 

I think the way to do is to change the Jenkinsfile-deploy so it does different things based on which stage we're on. It could go:
* If the stage is integration, carry on as it is now.
* If it's staging or prod check the code out from git, upload and update and just ignore the version returned. This means that the git version won't match the lambda version for staging and prod but it will always match with integration but we can create an alias against that version showing the git version, or we can compare the sha of the code, either way.

This means that we can deploy any particular version to any environment and, either by using aliases or the code sha, find out which version that was. 

That's all I can think of. What do you reckon?